src/swarm-scenario/Swarm/Game/Entity.hs

```diff
@@ -429,16 +430,36 @@ data EntityMap = EntityMap
 --   Note that duplicates in a single 'EntityMap' are precluded by the
 --   'buildEntityMap' function.  But it is possible for the right-hand
 --   'EntityMap' to override members of the left-hand with the same name.
---   This replacement happens automatically with 'Map', but needs to
---   be explicitly handled for the list concatenation of
---   'entityDefinitionOrder' (overridden entries are removed from the
---   former 'EntityMap').
+--   For example, this is how custom entities defined in a scenario
+--   can override standard entities. This replacement happens
+--   automatically with 'Map' (as long as we keep in mind that Map
+--   union is *left*-biased), but needs to be explicitly handled for the
+--   list concatenation of 'entityDefinitionOrder' (overridden entries
+--   are removed from the former 'EntityMap'), and for 'entitiesByCap',
+--   which are organized by capability rather than by entity.
 instance Semigroup EntityMap where
   EntityMap n1 c1 d1 <> EntityMap n2 c2 d2 =
     EntityMap
       (n2 <> n1)
-      (c1 <> c2)
-      (filter ((`M.notMember` n2) . view entityName) d1 <> d2)
+      (removeOverriddenDevices c1 <> c2)
+      (filter notOverridden d1 <> d2)
+   where
+    notOverridden :: Entity -> Bool
+    notOverridden = (`M.notMember` n2) . view entityName
+    removeOverriddenDevices (Capabilities m) =
+      Capabilities
+        . runIdentity
+        -- traverseMaybeWithKey allows us in one operation to filter
+        -- the NonEmpty list of devices for each capability, turn the
+        -- resulting lists into Maybe NonEmpty, and then remove from
+        -- the map any that became Nothing.  However, it is more
+        -- general than we need: it gives us access to the key which
+        -- we don't need (this explains the call to 'const'), and runs
+        -- in an arbitrary Applicative which we also don't need (this
+        -- explains the Identity).
+        . M.traverseMaybeWithKey
+          (const (Identity . NE.nonEmpty . NE.filter (notOverridden . device)))
```

Comment:
Instead of placing this verbose explanation inline here, shall we expose a simplified version of `traverseMaybeWithKey` in a util class?  Perhaps call the function `filterMaybes`?

Reply:
That's a good idea, I'll do that.

Reply:
Oh! It turns out it already exists in `Data.Map`, it's called `mapMaybeWithKey`.  Don't know why I didn't find that one before.